p2p/service.py

```diff
@@ -213,6 +216,15 @@ def run_daemon(self, service: 'BaseService') -> None:
 
         self.run_task(_run_daemon_wrapper())
 
+    def call_later(self, delay: float, callback: 'Callable[..., None]', *args: Any) -> None:
+
+        @functools.wraps(callback)
+        async def _call_later_wrapped() -> None:
+            self.sleep(delay)
```

Comment:
Is there a reason why you are not using the [`AbstractEventLoop.call_later`](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.AbstractEventLoop.call_later) API?

Reply:
Reason I ask is that if we use that, we shouldn't have to wrap the function if we use that API, though we will need to add another container to store the `TimeHandle` return values and cancel them upon cancellation.

Reply:
> we will need to add another container to store the TimeHandle return values and cancel them upon cancellation.

Yup, this is why. I didn't want to add yet another set that we had to remember to update in all the places. I started writing a `WeakSet[asyncio.TimerHandle]` and it just felt wrong to keep adding complexity to `BaseService` as it's already getting heavy. We can't treat it like the other tasks, and would explicitly have to cancel them right away at cancel time, which is fairly different from how the other tasks are dealt with (since they all usually die on their own with the cancel token).

Reply:
I guess it seems like a relatively simple addition, though I agree our `BaseService` API has grown substantially.

One way we might be able to battle some of that complexity would be beefing up the test suite, and looking into extracting into a dedicated library.  Forces us to document it, and think about breaking API changes which we currently don't do.